### DISCUSSION TITLE
Removal of shuttered APIs

### DISCUSSION THREAD

**Discussion by MikeRalphson** [APIs-guru/openapi-directory#1110] [Polls]
Following on from @andypiper's comment on issue #1052 - should we remove the definitions of shuttered APIs?

I see a few pros and cons.

* It would reduce the risk of pointing developers at APIs they can no longer use and thus maybe wasting their time
* It would enable developers to find actively developed APIs quicker
* Accessing old API definitions would become more complex for researchers / those comparing API changes / those wanting to re-implement an API (i.e. old git commits would have to be searched)
* The update logs would be less full of 404 errors
* If we are not informed, how long between an API url 404ing and removal?
* What if the API definition URL has simply changed and we don't know. Searching for moved URLs takes time

Another option would be to move old/shuttered/defunct/404ing APIs to another directory within the repo, this would hide them from the UI and the API but keep the files around for research and potential moving back to "live".

Would love to hear some thoughts.

**Comment by andypiper** [comment]
I've also been away for a few months (!) and I'm no longer involved at Twitter. It seems as though "advertising" shuttered or closed APIs via this project is only going to lead to disappointment for anyone wanting to use them, so best to remove, in my opinion.

**Reply by MikeRalphson** [reply]
Hi @andypiper - random question, but I see from your RSVP you might be at Oxford Python this evening?

**Reply by andypiper** [reply]
That's the plan, about to drive that way now - see you later!

**Comment by pimterry** [comment]
> old git commits would have to be searched

I think that's fine - well within what anybody digging into outdated API data is going to expect, and convenient enough for everybody involved. It's also very easy to bring things back from Git history, which I think negates the value of any 'recycle bin' approach.

> What if the API definition URL has simply changed and we don't know. Searching for moved URLs takes time

This is the hardest part I think... It would be really useful to get a feel for the typical reasons why a spec disappears to know how much this matters and how it's normally handled. Is there a list of the current 404s easily available somewhere? It would be interesting to do a quick sample of a few at random, and see if any obvious patterns jump out immediately.

There are some cases where we could be more nuanced though, for example in the GitHub-hosted case at least, we can handle 'GitHub temporarily unavailable' vs 'repo deleted' vs 'spec file deleted' all separately. Also in official vs unofficial spec cases - if an unofficial spec is removed, to me that sounds much less likely that the API is completely gone, and so maybe we should keep hosting it ourselves, whereas if the official spec is removed then something really has changed on the provider side (either a real update or a removal entirely) and we should be doing _something_ to handle it.

**Reply by MikeRalphson** [reply]
All your points are well made.

> Is there a list of the current 404s easily available somewhere?

@pimterry there is a report, I've dumped it into a [gist](https://gist.github.com/MikeRalphson/0b8a089cd7464deabbcd289593be01fd)

**Reply by pimterry** [reply]
I did a quick survey, just taking the bottom chunk of the gist (seemed as good a way to sample as any):

https://developer.tomtom.com/system/files/swagger_models/maps_api_0.yaml
https://developer.tomtom.com/system/files/swagger_models/routing_api.yaml
https://developer.tomtom.com/system/files/swagger_models/search_api.yaml
  - Spec gone - These 3 APIs still exist, seemingly unchanged and still using OpenAPI (see https://developer.tomtom.com/routing-api/api-explorer) but they don't obviously expose the specs themselves anywhere I can see

https://developer.transavia.com/docs/services/58d8bca5a9e6241bac7e89d8/export?DocumentFormat=Swagger
  - Spec gone - API exists, but looks like details (and presumably spec?) are only available with an API account (https://developer.transavia.com/docs/services/)

https://api.twitter.com/labs/1/openapi.json
  - API gone - was moved to https://api.twitter.com/1/openapi.json but that now returns a message about upgrading

https://api.twitter.com/labs/2/openapi.json
  - Spec moved - still exists, but now at https://api.twitter.com/2/openapi.json

https://vaccinationstatus.azurewebsites.net/covid-vaccine-status/v1.0.0/usa/api/v2/api-docs?group=athena-api
https://vaccinationstatus.azurewebsites.net/covid-vaccine-status/v1.0.0/brasil/api/v2/api-docs?group=brasil-api
https://vaccinationstatus.azurewebsites.net/covid-vaccine-status/v1.0.0/india/api/v2/api-docs?group=india-api
https://vaccinationstatus.azurewebsites.net/covid-vaccine-status/v1.0.0/paraguay/api/v2/api-docs?group=Paraguay-api
https://vaccinationstatus.azurewebsites.net/covid-vaccine-status/v1.0.0/peru/api/v2/api-docs?group=Peru-api
  - API gone (for all 5 of these - the domain no longer exists, was a temporary service during COVID)

https://vatapi.com/api-docs.json
  - Spec gone - company continues, but v1 API that used Swagger is now deprecated (but still works) and new v2 API doesn't seem to use OpenAPI explicitly

https://www.versioneye.com/api/v2/swagger_doc.json
  - Spec moved, now https://www.versioneye.com/api-docs/v1/swagger.yaml

https://vocadb.net/swagger/docs/v1
  - Spec moved, now https://vocadb.net/swagger/v1/swagger.json

https://api.citysdk.waag.org/swagger
  - API gone (short-lived pilot project, now finished: https://waag.org/en/project/smart-citysdk/)

https://watchful.li/api/v1/api-docs/
  - Spec split, now available as separate specs from https://app.watchful.net/apidoc-v1/ (raw -> spec in each case)

https://api.wavecell.com/swagger/v1/swagger.json
  - API gone (wavecell bought by 8x8: https://wavecell.com/8x8-acquisition/)

https://wealthport.github.io/wealthport-api-docs/swagger.json
  - API gone (spec references wealthport.com which no longer exists, GitHub wealthport account is gone too)

http://api.wedpax.com/swagger/v1/swagger.json
  - API gone (can't find any info, but the domain no longer exists)

https://sandbox.cloud.wowza.com/en/docs/api/v1
  - Spec no longer available. API still exists, but it's for a local self-hosted device, and the device now exposes its own spec directly (localhost:8089/swagger.json) and they don't publish it elsewhere. Interesting versioning solution!

https://raw.githubusercontent.com/XeroAPI/Xero-OpenAPI/master/spec/accounting2.yaml
  - Spec split (as reported in https://github.com/APIs-guru/openapi-directory/issues/1116)

https://yunbi.com/api/v2/doc/swagger
  - API gone (https://www.coindesk.com/markets/2017/09/15/yunbi-bitcoin-exchange-latest-to-close-in-china-crackdown/)

https://api.zalando.com/schema/swagger.json
  - Spec split, now available via https://developers.merchants.zalando.com/docs/api-ref-intro.html, with separate downloads for each API e.g. https://developers.merchants.zalando.com/docs/openapi/specs/products.yaml

http://zappiti.com/api/zappiti-player-4k/swagger/swagger.yaml
  - Spec gone? Swagger-based docs available at https://www.hdland.fr/swagger/ (linked from https://zappiti.uservoice.com/knowledgebase/articles/1831321-zappiti-video-http-service-on-player) but barely documented and raw spec doesn't seem available anywhere.

https://marketplace.zoom.us/docs/api-reference/zoom-api/Zoom%20API.oas2.json
  - Spec moved, now available from a data:base64 link button(???!!!) on https://marketplace.zoom.us/docs/api-reference/zoom-api/methods/#overview

---

Tallying up, that's:

* 5 specs now totally unavailable (but the APIs seem to continue working)
* 7 APIs that are completely gone
* 4 specs that are totally fine but the URL has changed
* 3 where a single spec has been replaced by multiple separate specs

(Counting TomTom & the Vax status site as one API, in each case)

**Reply by pimterry** [reply]
Having dug through, as a good heuristic to handle these I'd suggest trying to connect to the hosting domain & each of the server APIs, and if both fail persistently (for a month, say) then imo it's 99++++% chance that the API is gone forever and we should delete the spec. That would easily filter out a good chunk of the examples (1/3, in this sample) immediately.

Checking if the API servers themselves are still connectable would give you a reasonable-ish guess of whether the API is still actually up in reality (in which case we shouldn't drop the spec, imo) though it doesn't help with finding the correct new spec.

Oh, any server hostname checks will probably need to ignore anything that's localhost/*.local/IPs etc, to avoid marking self-hosted/local network APIs as dead.

Unfortunately, I think everything except 'completely dead' will still need manual intervention... At least this should result in a much shorter list, that may be possible to handle manually. Could we automatically create a new GitHub issue to track each case like that as new ones appear (once they've been down some reasonable duration) and then people can help out with fixing that? It usually only takes a minute or two to find out what's going on in each case. Also, that would provide a chance to quickly nag at companies that remove their OpenAPI specs, which might make them more likely to keep publishing them :smile:

**Reply by MikeRalphson** [reply]
@pimterry thanks for all the work that's gone into the above, both the concrete information and your suggestions going forward.

**Reply by MikeRalphson** [reply]
@pimterry I need to publish some proper development/contributing guidelines (see #1109) once I've taken a view on whether to open-source the registry package which now drives APIs.guru (it's currently a private repo/package on Github). There was a thought to allowing other interested parties to set up their own compatible repositories of internal and external API dependencies, but we never got any interest shown. There was also a consideration of what should be non-open in case someone wanted to acquire APIs.guru, but that seems a remote likelihood at the moment.

Opening things up would mean the registry metadata (currently a 3Mb blob of YAML, but I'm looking at [Acebase](https://github.com/appy-one/acebase)) could move into the main repo too. The new API submission automation scripts will probably move into Github workflows when they're more battle-tested.

For the cases where the API still exists in a 1:1 form and the URL just needs updating, would you be able to raise some PRs against the `info.x-origin.url` properties in the main repo YAML files? You don't need to amend the `format` or `version` properties or rename the files, that'll get done by an update sweep.

I feel a blog post coming on where we could shame API providers who break OpenAPI URLs over time!

**Reply by MikeRalphson** [reply]
I have a solution for `data:base64` and `blob://` URL definitions (as used by ReDoc) so that one will be ok.

**Reply by pimterry** [reply]
> There was a thought to allowing other interested parties to set up their own compatible repositories of internal and external API dependencies, but we never got any interest shown. There was also a consideration of what should be non-open in case someone wanted to acquire APIs.guru, but that seems a remote likelihood at the moment.

I have a sort-of similar situation for HTTP Toolkit - my conclusion so far is that licensing as AGPL with a tiny CLA for contributors (https://cla-assistant.io/httptoolkit/) works pretty well, and lets you go fully open-source without getting pinned into a corner forever. AGPL blocks any competition from closed-source direct forks of everything, and the CLA leaves you the option to change the license in future as required (because it explicitly gives you permission to use any license you like for the contributions you receive).

You can never retroactively stop people using the open source version once it's out, but this would allow you/somebody buying APIs.guru to say "from now on all future changes are open-source/closed source" or "yes, you can buy a non-GPL version for self-hosting, we will relicense it to you under a custom agreement for $$$" or similar.

I'm definitely not a lawyer, but it seems like a pretty tidy option AFAICT.

> For the cases where the API still exists in a 1:1 form and the URL just needs updating, would you be able to raise some PRs against the info.x-origin.url properties in the main repo YAML files?

Sure, I'll take a look at that tomorrow :+1:

I'm not totally clear what you mean about the `data:base64` solution - what should go in the origin in that case? The entire base64 blob???

**Reply by MikeRalphson** [reply]
It's slightly like something out of 1984 (we have always been at war with Eurasia) but I've used git filter-branch to retroactively license https://github.com/Mermade/api-registry as AGPL, added a CLA workflow, and open-sourced it. Thanks for the prompt.

**Reply by pimterry** [reply]
Awesome, great to hear, thanks @MikeRalphson! It'll be great to be able to dig into that directly next time I run into issues, that's super helpful :+1:
